File: backend/src/services/bedrock_service.py
import json
import boto3
import logging
from typing import Dict, List, Optional, Any
from botocore.exceptions import ClientError, BotoCoreError

logger = logging.getLogger(__name__)

class BedrockService:
    """
    Serviço para integração com Amazon Bedrock
    Fornece análise de IA para imóveis e oportunidades de investimento
    """
    
    def __init__(self, region_name: str = 'us-east-1'):
        """
        Inicializa o serviço Bedrock
        
        Args:
            region_name: Região AWS para o Bedrock
        """
        try:
            self.bedrock_runtime = boto3.client(
                service_name='bedrock-runtime',
                region_name=region_name
            )
            self.model_id = 'anthropic.claude-3-sonnet-20240229-v1:0'
        except Exception as e:
            logger.error(f"Failed to initialize Bedrock client: {e}")
            raise
    
    def analyze_property_opportunity(self, property_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Analisa uma oportunidade de investimento imobiliário usando IA
        
        Args:
            property_data: Dados do imóvel para análise
            
        Returns:
            Dict com análise detalhada da oportunidade
        """
        prompt = self._build_property_analysis_prompt(property_data)
        
        try:
            response = self._invoke_claude(prompt)
            return self._parse_property_analysis_response(response)
        except Exception as e:
            logger.error(f"Property analysis failed: {e}")
            return self._get_fallback_analysis()
    
    def generate_market_insights(self, market_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Gera insights de mercado baseados em dados agregados
        
        Args:
            market_data: Dados de mercado para análise
            
        Returns:
            Dict com insights e tendências de mercado
        """
        prompt = self._build_market_insights_prompt(market_data)
        
        try:
            response = self._invoke_claude(prompt)
            return self._parse_market_insights_response(response)
        except Exception as e:
            logger.error(f"Market insights generation failed: {e}")
            return self._get_fallback_market_insights()
    
    def analyze_investment_portfolio(self, portfolio_data: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Analisa um portfólio de investimentos imobiliários
        
        Args:
            portfolio_data: Lista de imóveis no portfólio
            
        Returns:
            Dict com análise do portfólio
        """
        prompt = self._build_portfolio_analysis_prompt(portfolio_data)
        
        try:
            response = self._invoke_claude(prompt)
            return self._parse_portfolio_analysis_response(response)
        except Exception as e:
            logger.error(f"Portfolio analysis failed: {e}")
            return self._get_fallback_portfolio_analysis()
    
    def generate_auction_strategy(self, property_data: Dict[str, Any], user_profile: Dict[str, Any]) -> Dict[str, Any]:
        """
        Gera estratégia personalizada para leilão
        
        Args:
            property_data: Dados do imóvel em leilão
            user_profile: Perfil do investidor
            
        Returns:
            Dict com estratégia de leilão personalizada
        """
        prompt = self._build_auction_strategy_prompt(property_data, user_profile)
        
        try:
            response = self._invoke_claude(prompt)
            return self._parse_auction_strategy_response(response)
        except Exception as e:
            logger.error(f"Auction strategy generation failed: {e}")
            return self._get_fallback_auction_strategy()
    
    def _invoke_claude(self, prompt: str, max_tokens: int = 2000) -> str:
        """
        Invoca o modelo Claude via Bedrock
        
        Args:
            prompt: Prompt para o modelo
            max_tokens: Número máximo de tokens na resposta
            
        Returns:
            Resposta do modelo
        """
        body = {
            "anthropic_version": "bedrock-2023-05-31",
            "max_tokens": max_tokens,
            "messages": [
                {
                    "role": "user",
                    "content": prompt
                }
            ],
            "temperature": 0.1,
            "top_p": 0.9
        }
        
        try:
            response = self.bedrock_runtime.invoke_model(
                modelId=self.model_id,
                body=json.dumps(body),
                contentType='application/json'
            )
            
            response_body = json.loads(response['body'].read())
            return response_body['content'][0]['text']
            
        except ClientError as e:
            logger.error(f"Bedrock API error: {e}")
            raise
        except Exception as e:
            logger.error(f"Unexpected error invoking Claude: {e}")
            raise
    
    def _build_property_analysis_prompt(self, property_data: Dict[str, Any]) -> str:
        """Constrói prompt para análise de imóvel"""
        return f"""
Você é um especialista em investimentos imobiliários no Brasil. Analise esta oportunidade de investimento em leilão da Caixa Econômica Federal e forneça uma análise detalhada.

DADOS DO IMÓVEL:
- Código: {property_data.get('codigo', 'N/A')}
- Tipo: {property_data.get('tipo_imovel', 'N/A')}
- Endereço: {property_data.get('endereco_completo', 'N/A')}
- Cidade: {property_data.get('cidade', 'N/A')}, {property_data.get('uf', 'N/A')}
- Área Total: {property_data.get('area_total', 'N/A')} m²
- Valor de Avaliação: R$ {property_data.get('valor_avaliacao', 0):,.2f}
- Valor de Venda: R$ {property_data.get('valor_venda', 0):,.2f}
- Desconto: {property_data.get('desconto_percentual', 0):.1f}%
- Aceita Financiamento: {'Sim' if property_data.get('aceita_financiamento') else 'Não'}
- Situação: {property_data.get('situacao_ocupacao', 'N/A')}

ANÁLISE SOLICITADA:
1. Score de Oportunidade (0-100) com justificativa
2. Principais pontos positivos (máximo 5)
3. Principais riscos e pontos de atenção (máximo 5)
4. Estratégia recomendada (locação, revenda, reforma)
5. Faixa de ROI esperado (anual)
6. Perfil de investidor recomendado
7. Dicas específicas para este imóvel

Responda em formato JSON estruturado:
{{
    "score": número,
    "recomendacao": "forte_compra|compra|neutro|cautela|evitar",
    "pontos_positivos": ["ponto1", "ponto2", ...],
    "riscos": ["risco1", "risco2", ...],
    "estrategia_recomendada": "texto",
    "roi_estimado": "X% a Y%",
    "perfil_investidor": "conservador|moderado|agressivo",
    "dicas_especificas": ["dica1", "dica2", ...]
}}
"""
    
    def _build_market_insights_prompt(self, market_data: Dict[str, Any]) -> str:
        """Constrói prompt para insights de mercado"""
        return f"""
Você é um analista de mercado imobiliário brasileiro. Analise os dados agregados e forneça insights sobre tendências e oportunidades.

DADOS DE MERCADO:
- Total de imóveis analisados: {market_data.get('total_properties', 0)}
- Preço médio: R$ {market_data.get('average_price', 0):,.2f}
- Desconto médio: {market_data.get('average_discount', 0):.1f}%
- Cidades principais: {', '.join(market_data.get('top_cities', []))}
- Tipos mais comuns: {', '.join(market_data.get('common_types', []))}
- Período de análise: {market_data.get('period', 'N/A')}

Forneça insights em formato JSON:
{{
    "tendencias_gerais": ["tendência1", "tendência2", ...],
    "oportunidades_destaque": ["oportunidade1", "oportunidade2", ...],
    "alertas_mercado": ["alerta1", "alerta2", ...],
    "recomendacoes_estrategicas": ["rec1", "rec2", ...],
    "previsoes": {{
        "proximo_trimestre": "texto",
        "proximo_ano": "texto"
    }}
}}
"""
    
    def _build_portfolio_analysis_prompt(self, portfolio_data: List[Dict[str, Any]]) -> str:
        """Constrói prompt para análise de portfólio"""
        portfolio_summary = {
            'total_properties': len(portfolio_data),
            'total_value': sum(p.get('valor_venda', 0) for p in portfolio_data),
            'cities': list(set(p.get('cidade', '') for p in portfolio_data)),
            'types': list(set(p.get('tipo_imovel', '') for p in portfolio_data))
        }
        
        return f"""
Analise este portfólio de investimentos imobiliários e forneça recomendações de otimização.

RESUMO DO PORTFÓLIO:
- Número de imóveis: {portfolio_summary['total_properties']}
- Valor total: R$ {portfolio_summary['total_value']:,.2f}
- Cidades: {', '.join(portfolio_summary['cities'])}
- Tipos de imóveis: {', '.join(portfolio_summary['types'])}

Responda em formato JSON:
{{
    "score_diversificacao": número,
    "pontos_fortes": ["ponto1", "ponto2", ...],
    "areas_melhoria": ["area1", "area2", ...],
    "recomendacoes_otimizacao": ["rec1", "rec2", ...],
    "risco_concentracao": "baixo|medio|alto",
    "estrategia_expansao": "texto"
}}
"""
    
    def _build_auction_strategy_prompt(self, property_data: Dict[str, Any], user_profile: Dict[str, Any]) -> str:
        """Constrói prompt para estratégia de leilão"""
        return f"""
Crie uma estratégia personalizada de leilão para este investidor.

DADOS DO IMÓVEL:
- Valor atual: R$ {property_data.get('valor_venda', 0):,.2f}
- Desconto: {property_data.get('desconto_percentual', 0):.1f}%
- Localização: {property_data.get('cidade', 'N/A')}

PERFIL DO INVESTIDOR:
- Experiência: {user_profile.get('experience_level', 'N/A')}
- Capital disponível: R$ {user_profile.get('available_capital', 0):,.2f}
- Objetivo: {user_profile.get('investment_goal', 'N/A')}
- Tolerância a risco: {user_profile.get('risk_tolerance', 'N/A')}

Forneça estratégia em JSON:
{{
    "lance_maximo_recomendado": número,
    "estrategia_lance": "texto",
    "pontos_atencao": ["ponto1", "ponto2", ...],
    "preparacao_necessaria": ["prep1", "prep2", ...],
    "plano_pos_arrematacao": "texto"
}}
"""
    
    def _parse_property_analysis_response(self, response: str) -> Dict[str, Any]:
        """Parseia resposta da análise de imóvel"""
        try:
            # Extrai JSON da resposta
            start = response.find('{')
            end = response.rfind('}') + 1
            if start != -1 and end != 0:
                json_str = response[start:end]
                return json.loads(json_str)
        except Exception as e:
            logger.error(f"Failed to parse property analysis response: {e}")
        
        return self._get_fallback_analysis()
    
    def _parse_market_insights_response(self, response: str) -> Dict[str, Any]:
        """Parseia resposta dos insights de mercado"""
        try:
            start = response.find('{')
            end = response.rfind('}') + 1
            if start != -1 and end != 0:
                json_str = response[start:end]
                return json.loads(json_str)
        except Exception as e:
            logger.error(f"Failed to parse market insights response: {e}")
        
        return self._get_fallback_market_insights()
    
    def _parse_portfolio_analysis_response(self, response: str) -> Dict[str, Any]:
        """Parseia resposta da análise de portfólio"""
        try:
            start = response.find('{')
            end = response.rfind('}') + 1
            if start != -1 and end != 0:
                json_str = response[start:end]
                return json.loads(json_str)
        except Exception as e:
            logger.error(f"Failed to parse portfolio analysis response: {e}")
        
        return self._get_fallback_portfolio_analysis()
    
    def _parse_auction_strategy_response(self, response: str) -> Dict[str, Any]:
        """Parseia resposta da estratégia de leilão"""
        try:
            start = response.find('{')
            end = response.rfind('}') + 1
            if start != -1 and end != 0:
                json_str = response[start:end]
                return json.loads(json_str)
        except Exception as e:
            logger.error(f"Failed to parse auction strategy response: {e}")
        
        return self._get_fallback_auction_strategy()
    
    def _get_fallback_analysis(self) -> Dict[str, Any]:
        """Retorna análise de fallback quando a IA falha"""
        return {
            "score": 50,
            "recomendacao": "neutro",
            "pontos_positivos": ["Análise detalhada temporariamente indisponível"],
            "riscos": ["Recomenda-se análise manual detalhada"],
            "estrategia_recomendada": "Consulte um especialista para análise completa",
            "roi_estimado": "A definir",
            "perfil_investidor": "moderado",
            "dicas_especificas": ["Realize due diligence completa antes de investir"]
        }
    
    def _get_fallback_market_insights(self) -> Dict[str, Any]:
        """Retorna insights de fallback"""
        return {
            "tendencias_gerais": ["Análise de mercado temporariamente indisponível"],
            "oportunidades_destaque": ["Consulte dados de mercado atualizados"],
            "alertas_mercado": ["Mantenha-se atualizado com indicadores econômicos"],
            "recomendacoes_estrategicas": ["Diversifique investimentos"],
            "previsoes": {
                "proximo_trimestre": "Dados insuficientes para previsão",
                "proximo_ano": "Acompanhe indicadores macroeconômicos"
            }
        }
    
    def _get_fallback_portfolio_analysis(self) -> Dict[str, Any]:
        """Retorna análise de portfólio de fallback"""
        return {
            "score_diversificacao": 50,
            "pontos_fortes": ["Análise detalhada em processamento"],
            "areas_melhoria": ["Consulte especialista para análise completa"],
            "recomendacoes_otimizacao": ["Revise periodicamente seu portfólio"],
            "risco_concentracao": "medio",
            "estrategia_expansao": "Análise personalizada recomendada"
        }
    
    def _get_fallback_auction_strategy(self) -> Dict[str, Any]:
        """Retorna estratégia de leilão de fallback"""
        return {
            "lance_maximo_recomendado": 0,
            "estrategia_lance": "Consulte especialista para estratégia personalizada",
            "pontos_atencao": ["Análise de risco necessária", "Verificar documentação"],
            "preparacao_necessaria": ["Due diligence completa", "Análise financeira"],
            "plano_pos_arrematacao": "Plano personalizado recomendado"
        }

File: backend/app/Services/BedrockService.php
<?php

namespace App\Services;

use Aws\BedrockRuntime\BedrockRuntimeClient;
use Aws\Exception\AwsException;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Facades\Cache;

class BedrockService
{
    private $client;
    private $modelId;
    private $maxTokens;
    private $temperature;
    
    public function __construct()
    {
        $this->client = new BedrockRuntimeClient([
            'region' => config('services.aws.region', 'us-east-1'),
            'version' => 'latest',
            'credentials' => [
                'key' => config('services.aws.key'),
                'secret' => config('services.aws.secret'),
            ],
        ]);
        
        $this->modelId = config('services.bedrock.model_id', 'anthropic.claude-3-haiku-20240307-v1:0');
        $this->maxTokens = config('services.bedrock.max_tokens', 4000);
        $this->temperature = config('services.bedrock.temperature', 0.1);
    }

    /**
     * Extrai dados de propriedade usando IA
     */
    public function extractPropertyData($htmlContent)
    {
        $cacheKey = 'bedrock_extraction_' . md5($htmlContent);
        
        // Verificar cache para evitar custos desnecessários
        if (Cache::has($cacheKey)) {
            return Cache::get($cacheKey);
        }
        
        $prompt = $this->buildPropertyExtractionPrompt($htmlContent);
        
        try {
            $response = $this->invokeModel($prompt);
            $extractedData = $this->parsePropertyResponse($response);
            
            // Cache por 24 horas
            Cache::put($cacheKey, $extractedData, now()->addHours(24));
            
            return $extractedData;
            
        } catch (\Exception $e) {
            Log::error('Erro na extração Bedrock: ' . $e->getMessage());
            throw $e;
        }
    }

    /**
     * Analisa mudanças na estrutura do site da Caixa
     */
    public function analyzeStructureChanges($currentHtml, $previousHtml = null)
    {
        if (!$previousHtml) {
            $previousHtml = Cache::get('last_caixa_structure', '');
        }
        
        if (empty($previousHtml)) {
            Cache::put('last_caixa_structure', $currentHtml, now()->addDays(7));
            return ['changes_detected' => false];
        }
        
        $prompt = $this->buildStructureAnalysisPrompt($currentHtml, $previousHtml);
        
        try {
            $response = $this->invokeModel($prompt);
            $analysis = $this->parseStructureAnalysis($response);
            
            if ($analysis['changes_detected']) {
                // Atualizar estrutura de referência
                Cache::put('last_caixa_structure', $currentHtml, now()->addDays(7));
                
                // Notificar sobre mudanças
                $this->notifyStructureChanges($analysis);
            }
            
            return $analysis;
            
        } catch (\Exception $e) {
            Log::error('Erro na análise de estrutura: ' . $e->getMessage());
            return ['changes_detected' => false, 'error' => $e->getMessage()];
        }
    }

    /**
     * Gera novos seletores CSS/XPath baseado em mudanças
     */
    public function generateNewSelectors($htmlContent, $targetData)
    {
        $prompt = $this->buildSelectorGenerationPrompt($htmlContent, $targetData);
        
        try {
            $response = $this->invokeModel($prompt);
            return $this->parseSelectorResponse($response);
            
        } catch (\Exception $e) {
            Log::error('Erro na geração de seletores: ' . $e->getMessage());
            return [];
        }
    }

    /**
     * Analisa oportunidades de investimento
     */
    public function analyzeInvestmentOpportunity($propertyData, $marketData = [])
    {
        $prompt = $this->buildInvestmentAnalysisPrompt($propertyData, $marketData);
        
        try {
            $response = $this->invokeModel($prompt);
            return $this->parseInvestmentAnalysis($response);
            
        } catch (\Exception $e) {
            Log::error('Erro na análise de investimento: ' . $e->getMessage());
            return null;
        }
    }

    /**
     * Gera insights sobre tendências do mercado
     */
    public function generateMarketInsights($properties, $timeframe = '30d')
    {
        $prompt = $this->buildMarketInsightsPrompt($properties, $timeframe);
        
        try {
            $response = $this->invokeModel($prompt);
            return $this->parseMarketInsights($response);
            
        } catch (\Exception $e) {
            Log::error('Erro na geração de insights: ' . $e->getMessage());
            return [];
        }
    }

    /**
     * Otimiza estratégias de leilão
     */
    public function optimizeAuctionStrategy($propertyData, $userProfile, $marketConditions)
    {
        $prompt = $this->buildAuctionStrategyPrompt($propertyData, $userProfile, $marketConditions);
        
        try {
            $response = $this->invokeModel($prompt);
            return $this->parseAuctionStrategy($response);
            
        } catch (\Exception $e) {
            Log::error('Erro na otimização de estratégia: ' . $e->getMessage());
            return null;
        }
    }

    /**
     * Invoca o modelo Bedrock
     */
    private function invokeModel($prompt, $systemPrompt = null)
    {
        $body = [
            'anthropic_version' => 'bedrock-2023-05-31',
            'max_tokens' => $this->maxTokens,
            'temperature' => $this->temperature,
            'messages' => [
                [
                    'role' => 'user',
                    'content' => $prompt
                ]
            ]
        ];
        
        if ($systemPrompt) {
            $body['system'] = $systemPrompt;
        }
        
        try {
            $result = $this->client->invokeModel([
                'modelId' => $this->modelId,
                'contentType' => 'application/json',
                'accept' => 'application/json',
                'body' => json_encode($body),
            ]);
            
            $response = json_decode($result['body']->getContents(), true);
            
            if (isset($response['content'][0]['text'])) {
                return $response['content'][0]['text'];
            }
            
            throw new \Exception('Resposta inválida do Bedrock');
            
        } catch (AwsException $e) {
            Log::error('Erro AWS Bedrock: ' . $e->getMessage());
            throw $e;
        }
    }

    /**
     * Constrói prompt para extração de dados de propriedade
     */
    private function buildPropertyExtractionPrompt($htmlContent)
    {
        return "
        Você é um especialista em extração de dados de imóveis. Analise o HTML abaixo de uma página da Caixa Econômica Federal e extraia as informações solicitadas.

        INSTRUÇÕES IMPORTANTES:
        1. Procure especificamente por informações sobre FINANCIAMENTO - isso é CRÍTICO
        2. Extraia valores monetários no formato correto
        3. Identifique modalidade de venda (leilão, venda direta, etc.)
        4. Se não encontrar uma informação, retorne null
        5. Retorne APENAS um JSON válido

        CAMPOS OBRIGATÓRIOS:
        - aceita_financiamento (boolean): true se aceita, false se não aceita, null se indeterminado
        - valor_venda (number): valor de venda em reais
        - valor_avaliacao (number): valor de avaliação em reais
        - modalidade_venda (string): tipo de venda
        - situacao_imovel (string): situação atual
        - ocupacao (string): ocupado/desocupado
        - endereco_completo (string): endereço completo
        - descricao (string): descrição do imóvel
        - matricula (string): número da matrícula
        - area_privativa (number): área em m²
        - quartos (number): número de quartos
        - banheiros (number): número de banheiros
        - vagas_garagem (number): vagas de garagem

        HTML PARA ANÁLISE:
        {$htmlContent}

        RESPOSTA (apenas JSON):
        ";
    }

    /**
     * Constrói prompt para análise de mudanças estruturais
     */
    private function buildStructureAnalysisPrompt($currentHtml, $previousHtml)
    {
        return "
        Compare as duas estruturas HTML abaixo e identifique se houve mudanças significativas na estrutura da página de detalhes de imóveis da Caixa.

        FOQUE EM:
        1. Mudanças nos seletores CSS/XPath para campos importantes
        2. Alterações na estrutura de tabelas ou divs
        3. Novos campos ou campos removidos
        4. Mudanças nos padrões de texto para financiamento

        HTML ANTERIOR:
        {$previousHtml}

        HTML ATUAL:
        {$currentHtml}

        Retorne um JSON com:
        {
            \"changes_detected\": boolean,
            \"severity\": \"low|medium|high\",
            \"affected_fields\": [\"campo1\", \"campo2\"],
            \"recommended_actions\": [\"ação1\", \"ação2\"],
            \"new_selectors_needed\": boolean
        }
        ";
    }

    /**
     * Constrói prompt para geração de seletores
     */
    private function buildSelectorGenerationPrompt($htmlContent, $targetData)
    {
        return "
        Analise o HTML e gere seletores CSS e XPath otimizados para extrair os dados especificados.

        DADOS ALVO: " . json_encode($targetData) . "

        HTML:
        {$htmlContent}

        Retorne JSON com seletores no formato:
        {
            \"campo\": {
                \"css\": \"seletor css\",
                \"xpath\": \"seletor xpath\",
                \"confidence\": 0.95
            }
        }
        ";
    }

    /**
     * Constrói prompt para análise de investimento
     */
    private function buildInvestmentAnalysisPrompt($propertyData, $marketData)
    {
        return "
        Analise esta oportunidade de investimento imobiliário considerando os dados fornecidos.

        DADOS DO IMÓVEL: " . json_encode($propertyData) . "
        DADOS DE MERCADO: " . json_encode($marketData) . "

        Forneça análise detalhada incluindo:
        1. Score de oportunidade (0-100)
        2. Potencial de valorização
        3. Riscos identificados
        4. Recomendações de lance
        5. ROI estimado
        6. Tempo estimado para venda/aluguel

        Retorne JSON estruturado com a análise.
        ";
    }

    /**
     * Constrói prompt para insights de mercado
     */
    private function buildMarketInsightsPrompt($properties, $timeframe)
    {
        return "
        Analise os dados de imóveis e gere insights de mercado para o período de {$timeframe}.

        DADOS: " . json_encode($properties) . "

        Gere insights sobre:
        1. Tendências de preços por região
        2. Tipos de imóveis mais disponíveis
        3. Oportunidades emergentes
        4. Padrões de financiamento
        5. Recomendações estratégicas

        Retorne JSON com insights estruturados.
        ";
    }

    /**
     * Constrói prompt para estratégia de leilão
     */
    private function buildAuctionStrategyPrompt($propertyData, $userProfile, $marketConditions)
    {
        return "
        Desenvolva uma estratégia otimizada de leilão para este imóvel.

        IMÓVEL: " . json_encode($propertyData) . "
        PERFIL DO USUÁRIO: " . json_encode($userProfile) . "
        CONDIÇÕES DE MERCADO: " . json_encode($marketConditions) . "

        Forneça estratégia incluindo:
        1. Lance inicial recomendado
        2. Lance máximo sugerido
        3. Timing de lances
        4. Análise da concorrência esperada
        5. Plano B se não conseguir o imóvel
        6. Checklist pré-leilão

        Retorne JSON estruturado.
        ";
    }

    /**
     * Processa resposta de extração de propriedade
     */
    private function parsePropertyResponse($response)
    {
        try {
            // Limpar resposta para extrair apenas o JSON
            $response = trim($response);
            
            // Procurar por JSON na resposta
            if (preg_match('/\{.*\}/s', $response, $matches)) {
                $json = json_decode($matches[0], true);
                
                if (json_last_error() === JSON_ERROR_NONE) {
                    return $json;
                }
            }
            
            // Fallback: tentar decodificar a resposta inteira
            return json_decode($response, true) ?: [];
            
        } catch (\Exception $e) {
            Log::error('Erro ao processar resposta de propriedade: ' . $e->getMessage());
            return [];
        }
    }

    /**
     * Processa análise de estrutura
     */
    private function parseStructureAnalysis($response)
    {
        try {
            $analysis = json_decode($response, true);
            
            return [
                'changes_detected' => $analysis['changes_detected'] ?? false,
                'severity' => $analysis['severity'] ?? 'low',
                'affected_fields' => $analysis['affected_fields'] ?? [],
                'recommended_actions' => $analysis['recommended_actions'] ?? [],
                'new_selectors_needed' => $analysis['new_selectors_needed'] ?? false,
            ];
            
        } catch (\Exception $e) {
            Log::error('Erro ao processar análise de estrutura: ' . $e->getMessage());
            return ['changes_detected' => false];
        }
    }

    /**
     * Processa resposta de seletores
     */
    private function parseSelectorResponse($response)
    {
        try {
            return json_decode($response, true) ?: [];
        } catch (\Exception $e) {
            Log::error('Erro ao processar seletores: ' . $e->getMessage());
            return [];
        }
    }

    /**
     * Processa análise de investimento
     */
    private function parseInvestmentAnalysis($response)
    {
        try {
            return json_decode($response, true);
        } catch (\Exception $e) {
            Log::error('Erro ao processar análise de investimento: ' . $e->getMessage());
            return null;
        }
    }

    /**
     * Processa insights de mercado
     */
    private function parseMarketInsights($response)
    {
        try {
            return json_decode($response, true) ?: [];
        } catch (\Exception $e) {
            Log::error('Erro ao processar insights: ' . $e->getMessage());
            return [];
        }
    }

    /**
     * Processa estratégia de leilão
     */
    private function parseAuctionStrategy($response)
    {
        try {
            return json_decode($response, true);
        } catch (\Exception $e) {
            Log::error('Erro ao processar estratégia: ' . $e->getMessage());
            return null;
        }
    }

    /**
     * Notifica sobre mudanças estruturais
     */
    private function notifyStructureChanges($analysis)
    {
        // Enviar notificação para administradores
        Log::warning('Mudanças detectadas na estrutura da Caixa', $analysis);
        
        // Aqui você pode implementar notificações por email, Slack, etc.
    }

    /**
     * Otimiza custos limitando uso da IA
     */
    public function shouldUseAI($context)
    {
        // Lógica para decidir quando usar IA baseado em:
        // - Falhas na extração padrão
        // - Importância do imóvel
        // - Budget disponível
        // - Horário (usar menos IA em horários de pico)
        
        $hour = now()->hour;
        $isBusinessHour = $hour >= 8 && $hour <= 18;
        
        // Usar menos IA durante horário comercial para otimizar custos
        if ($isBusinessHour && $context['priority'] !== 'high') {
            return rand(1, 100) <= 30; // 30% de chance
        }
        
        return true;
    }
}

